mpt_trie/src/utils.rs

```diff
@@ -58,3 +77,190 @@ pub(crate) fn create_mask_of_1s(amt: usize) -> U512 {
+        self.0.into_iter()
+    }
+}
+
+impl From<Vec<TrieSegment>> for TriePath {
+    fn from(v: Vec<TrieSegment>) -> Self {
+        Self(v)
+    }
+}
+
+impl FromIterator<TrieSegment> for TriePath {
+    fn from_iter<T: IntoIterator<Item = TrieSegment>>(iter: T) -> Self {
+        Self(Vec::from_iter(iter))
+    }
+}
+
+impl TriePath {
+    /// Get an iterator of the individual path segments in the [`TriePath`].
+    pub fn iter(&self) -> impl Iterator<Item = &'_ TrieSegment> {
+        self.0.iter()
+    }
+
+    pub(crate) fn dup_and_append(&self, seg: TrieSegment) -> Self {
+        let mut duped_vec = self.0.clone();
+        duped_vec.push(seg);
+
+        Self(duped_vec)
+    }
+
+    pub(crate) fn append(&mut self, seg: TrieSegment) {
+        self.0.push(seg);
+    }
+
+    fn write_elem(f: &mut fmt::Formatter<'_>, seg: &TrieSegment) -> fmt::Result {
+        write!(f, "{}", seg)
+    }
+}
+
+impl Display for TrieSegment {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            TrieSegment::Empty => write!(f, "Empty"),
+            TrieSegment::Hash => write!(f, "Hash"),
+            TrieSegment::Branch(nib) => write!(f, "Branch({})", nib),
+            TrieSegment::Extension(nibs) => write!(f, "Extension({})", nibs),
+            TrieSegment::Leaf(nibs) => write!(f, "Leaf({})", nibs),
+        }
+    }
+}
+
+impl TrieSegment {
+    /// Get the node type of the [`TrieSegment`].
+    pub fn node_type(&self) -> TrieNodeType {
+        match self {
+            TrieSegment::Empty => TrieNodeType::Empty,
+            TrieSegment::Hash => TrieNodeType::Hash,
+            TrieSegment::Branch(_) => TrieNodeType::Branch,
+            TrieSegment::Extension(_) => TrieNodeType::Extension,
+            TrieSegment::Leaf(_) => TrieNodeType::Leaf,
+        }
+    }
+
+    /// Extracts the key piece used by the segment (if applicable).
+    pub fn get_key_piece_from_seg_if_present(&self) -> Option<Nibbles> {
+        match self {
+            TrieSegment::Empty | TrieSegment::Hash => None,
+            TrieSegment::Branch(nib) => Some(Nibbles::from_nibble(*nib)),
+            TrieSegment::Extension(nibs) | TrieSegment::Leaf(nibs) => Some(*nibs),
+        }
+    }
+}
+
+/// Creates a [`TrieSegment`] given a node and a key we are querying.
+///
+/// This function is intended to be used during a trie query as we are
+/// traversing down a trie. Depending on the current node, we pop off nibbles
+/// and use these to create `TrieSegment`s.
+pub fn get_segment_from_node_and_key_piece<T: PartialTrie>(
+    n: &Node<T>,
+    k_piece: &Nibbles,
+) -> TrieSegment {
+    match TrieNodeType::from(n) {
+        TrieNodeType::Empty => TrieSegment::Empty,
+        TrieNodeType::Hash => TrieSegment::Hash,
+        TrieNodeType::Branch => TrieSegment::Branch(k_piece.get_nibble(0)),
+        TrieNodeType::Extension => TrieSegment::Extension(*k_piece),
+        TrieNodeType::Leaf => TrieSegment::Leaf(*k_piece),
+    }
+}
```

Comment:
On the fence a bit if these three should be `pub`.

Reply:
I'd probably remove them for now. If someone has a need in the future then we can consider it but for the time being let's keep things clean and simple and avoid handling hypothetical scenarios.